oak_functions_service/src/lookup_htbl.rs

```diff
@@ -0,0 +1,575 @@
+// cache misses per lookup should be close to 2 for values found in the table, and 1 for values not
+// in the table.
+//
+// The storage for k/v pairs is chunked into 256 4GiB vectors:
+//
+//     data_chunks: Vec<Box([u8; N])>,  // N = 2^32
+//     used_chunks: usize,
+//
+// data_chunks starts empty, and as data is added to the table, we allocat another chunk.
+// The format for each k/v pair of data in each chunk is:
+//
+//     key_len: <compressed u30>
+//     key: [u8]
+//     value_len: <compressed u30>
+//     value: [u8]
+//
+// Another memory savings comes from using a swiss table with exactly a 60% load factor.  Swiss
+// tables require a significant portion of the table be empty, and normally as the table grows,
+// they dynamically double the size.  At least when using the alloc crate, this resize causes a new
+// table to be allocated at 2X the size, data is copied from the old table to the new, and the old
+// table is freed.  This results in a 3X expansion in memory usage vs the prior array size.  In
+// contrast, this scheme has only 40% of the table empty and is never resized.
+//
+// This reduce scheme currently limits the number of entries to 2^32.
+
+const CHUNK_BITS: usize = 21;
+const CHUNK_SIZE: usize = 1 << CHUNK_BITS;
+const CHUNK_MASK: usize = CHUNK_SIZE - 1;
+
+#[derive(Default)]
+pub struct LookupHtbl {
+    table: Vec<u8>, // Contains fixed-size entries with index of k/v data in data_chunks.
+    data_chunks: Vec<Box<[u8]>>,
+    max_entries: usize,       // The number at which we must grow the table.
+    allocated_entries: usize, // This is self.table.len() * ENTRY_SIZE.
+    used_data: usize,
+    used_entries: usize,
+}
+
+enum LookupResult {
+    Found(usize),        // (data_index: u40)
+    NotFound(usize, u8), // (table_index: u40, hash_byte: u8)
+}
+
+impl LookupHtbl {
+    // Set the initial size of self.table.  For best speed and memory use, this should be as large
+    // as the number of key/value entries that will be loaded.
+    pub fn reserve(&mut self, max_entries: usize) {
+        assert!(self.used_entries == 0 && self.used_data == 0);
+        // Use a load factor of 60%.
+        let allocated_entries = (5 * max_entries + 1) / 3;
+        self.max_entries = max_entries;
+        self.table = vec![0u8; ENTRY_SIZE * allocated_entries];
+        self.allocated_entries = allocated_entries;
+        // Allocate 1 extra byte, so the 0 index can be used as NULL.
+        self.data_chunks = vec![];
+        self.used_data = 1usize;
+        self.used_entries = 0usize;
+    }
+
+    // Lookup the key/value pair.  If found, return LookupFound(data_index).  Otherwise, return the
+    // table index that contains 0 which can be used during insert, as NotFound(table_index).
+    fn lookup(&self, key: &[u8]) -> LookupResult {
+        if self.allocated_entries == 0 {
+            panic!("Don't call lookup on empty table");
```

Comment:
I definitely disagree with panicking on an empty table.

Reply:
The panic message isn't clear.  This non-public function requires the caller to have checked that the table is not empty before calling lookup.  Otherwise we we have to have self be mutable, so we could return a valid index in the NotFound enum.  I've updated the panic message to be more clear.